File: app/src/main/java/com/example/puyo_base_simulator/ui/home/Field.java
package com.example.puyo_base_simulator.ui.home;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.Stack;

public class Field implements Cloneable {
    Field nextField = null;
    Puyo[][] field;
    int[] heights = {0,0,0,0,0,0,0};
    List<Puyo> disappearPuyo = new ArrayList<>();
    int accumulatedPoint;
    int bonus;
    int colorBonus;
    int chainNum;
    int connectionBonus;
    final int[] chainBonusConstant = {0, 0, 8, 16, 32, 64, 96, 128, 160, 192, 224, 256, 288, 320, 352, 384, 416, 448, 480, 512};
    final int[] colorBonusConstant = {0, 0, 3, 6, 12};
    int connectionBonusConstant(int connectionNum) {
        if (connectionNum <= 4) {
            return 0;
        } else if (connectionNum <= 10){
            return connectionNum - 3;
        } else {
            return 10;
        }
    }

    @NonNull
    @Override
    public Field clone() {
        Field cloned = new Field(this.chainNum);
        try {
            cloned = (Field) super.clone();
            cloned.field = this.field.clone();
            for (int i=1; i<14; i++) {
                cloned.field[i] = this.field[i].clone();
            }
            cloned.heights = this.heights.clone();
        } catch (CloneNotSupportedException e) {
            e.printStackTrace();
        }
        return cloned;
    }

    Field (int chainNum) {
        field = new Puyo[14][7];
        for (int i=1; i<14; i++) {
            for (int j=1; j<7; j++) {
                field[i][j] = new Puyo(i, j, PuyoColor.EMPTY);
            }
        }
        this.chainNum = chainNum;
    }

    Boolean addPuyo(int column, PuyoColor color) {
        int row = heights[column] + 1;
        if (row == 14) return false;

        field[row][column] = new Puyo(row, column, color);
        heights[column]++;
        return true;
    }

    boolean allClear() {
        for (int h : this.heights) {
            if (h > 0) return false;
        }
        return true;
    }

    boolean isDisappear(Puyo puyo) {
        return disappearPuyo.contains(puyo);
    }

    void evalNextField() {
        Field newField = new Field(this.chainNum + 1);
        Set<PuyoColor> colors = new HashSet<>();
        int connectionBonus = 0;
        // 消えるぷよを探す
        for (int i=1; i<13; i++) {
            for (int j=1; j<7; j++) {
                Puyo puyo = field[i][j];
                List<Puyo> connection = getConnection(puyo);
                if (connection.size() >= 4) {
                    disappearPuyo.add(puyo);
                    // 色ボーナスの評価
                    colors.add(puyo.color);
                    // 連結ボーナスの評価
                    boolean connectionIsNew = true;
                    for (Puyo p: disappearPuyo) {
                        if (p.row < i || p.row == i && p.column < j) {
                            connectionIsNew = false;
                            break;
                        }
                    }
                    if (connectionIsNew) {
                        connectionBonus += connectionBonusConstant(connection.size());
                    }
                } else if (puyo.color != PuyoColor.EMPTY){
                    newField.addPuyo(j, puyo.color);
                }
            }
        }
        if (newField.allClear()) this.accumulatedPoint += 3000;
        if (disappearPuyo.size() == 0) {
            this.chainNum = 1;
            return;
        }

        // 消えるぷよがある場合のみ次の盤面を評価
        int bonus = colorBonusConstant[colors.size()] + connectionBonus + chainBonusConstant[chainNum];
        if (bonus == 0) bonus = 1;
        int point = accumulatedPoint + bonus * disappearPuyo.size() * 10;

        this.colorBonus = colorBonusConstant[colors.size()];
        this.connectionBonus = connectionBonus;
        this.bonus = bonus;
        this.accumulatedPoint = point;
        this.nextField = newField;
        this.nextField.accumulatedPoint = point;
        newField.evalNextField();
    }

    List<Puyo> getNeighborPuyo(Puyo puyo) {
        int row = puyo.row;
        int column = puyo.column;
        List<Puyo> ret = new ArrayList<>();
        // left
        if (column != 1 && field[row][column-1].color != PuyoColor.EMPTY) {
            ret.add(field[row][column-1]);
        }
        // right
        if (column != 6 && field[row][column+1].color != PuyoColor.EMPTY) {
            ret.add(field[row][column+1]);
        }
        // up
        if (row != 12 && field[row+1][column].color != PuyoColor.EMPTY) {
            ret.add(field[row+1][column]);
        }
        // down
        if (row != 1 && field[row-1][column].color != PuyoColor.EMPTY) {
            ret.add(field[row-1][column]);
        }
        return ret;
    }

    // 連結数
    List<Puyo> getConnection(Puyo puyo) {
        ArrayList<Puyo> connected = new ArrayList<>();
        if (puyo.color == PuyoColor.EMPTY) return connected;
        Stack<Puyo> sameColorStack = new Stack<>();
        sameColorStack.push(puyo);
        connected.add(puyo);
        while (!sameColorStack.isEmpty()) {
            Puyo currentPuyo = sameColorStack.pop();
            List<Puyo> neighbors = getNeighborPuyo(currentPuyo);
            for (Puyo p: neighbors) {
                if (p.color == puyo.color && !connected.contains(p)) {
                    sameColorStack.push(p);
                    connected.add(p);
                }
            }
        }
        return connected;
    }

    @NonNull
    public String toString() {
        StringBuilder str = new StringBuilder();
        for (int i=1; i<14; i++) {
            for (int j = 1; j < 7; j++) {
                str.append(field[i][j].color.getChar());
            }
        }
        return str.toString();
    }

    // fromString
    Field (String fieldStr) {
        int idx = 0;
        field = new Puyo[14][7];
        for (int i=1; i<14; i++) {
            for (int j=1; j<7; j++) {
                PuyoColor color = PuyoColor.getPuyoColor(fieldStr.charAt(idx));
                field[i][j] = new Puyo(i, j, color);
                if (color != PuyoColor.EMPTY) {
                    heights[j]++;
                }
                idx++;
            }
        }
        this.chainNum = 1;
    }

}

File: app/src/main/java/com/example/puyo_base_simulator/ui/home/Field.kt
package com.example.puyo_base_simulator.ui.home

import com.example.puyo_base_simulator.ui.home.PuyoColor
import com.example.puyo_base_simulator.ui.home.PuyoColor.Companion.getPuyoColor
import java.io.Serializable
import java.util.*

class Field : Serializable {
    @JvmField
    var nextField: Field? = null
    var field = Array(13) { i -> Array(6) { j -> Puyo(i+1, j+1, PuyoColor.EMPTY)} }
    private fun getFieldContent(row: Int, column: Int) : Puyo { return field[row-1][column-1]}
    private var heights = Array(6 ) { 0 }
    fun getHeight(column: Int) : Int { return heights[column-1]}
    @JvmField
    var disappearPuyo = mutableListOf<Puyo>()
    @JvmField
    var accumulatedPoint = 0
    @JvmField
    var bonus = 0
    private var chainNum = 0

    fun addPuyo(column: Int, color: PuyoColor): Boolean {
        val row = heights[column-1] + 1
        if (row == 14) return false
        field[row-1][column-1] = Puyo(row, column, color)
        heights[column-1]++
        return true
    }

    fun allClear() = heights.all {it == 0}

    fun isDisappear(puyo: Puyo) = disappearPuyo.contains(puyo)

    fun evalNextField() {
        val newField = Field()
        newField.chainNum = chainNum + 1
        var connectionBonus = 0
        // 消えるぷよを探す
        for (puyo in field.sliceArray(0..11).flatten()) {
            if (disappearPuyo.contains(puyo) || puyo.color == PuyoColor.EMPTY) continue
            val connection = getConnection(puyo)
            if (connection.size >= 4) {
                disappearPuyo.addAll(connection)
                connectionBonus += connectionBonusConstant(connection.size)
            } else {
                newField.addPuyo(puyo.column, puyo.color)
            }
        }
        if (newField.allClear()) accumulatedPoint += allClearBonus
        if (disappearPuyo.size == 0) {
            chainNum = 1
            return
        }

        // 消えるぷよがある場合のみ次の盤面を評価
        val colorNum = disappearPuyo.map {it.color}.toSet().size
        var bonus = colorBonusConstant[colorNum] + connectionBonus + chainBonusConstant[chainNum]
        if (bonus == 0) bonus = 1
        val point = accumulatedPoint + bonus * disappearPuyo.size * 10
        this.bonus = bonus
        accumulatedPoint = point
        nextField = newField
        nextField!!.accumulatedPoint = point
        newField.evalNextField()
    }

    private fun getNeighborPuyo(puyo: Puyo): List<Puyo> {
        val row = puyo.row
        val column = puyo.column
        val neighbors = mutableListOf<Puyo>()
        // left
        if (column != 1) {
            neighbors.add(getFieldContent(row,column - 1))
        }
        // right
        if (column != 6) {
            neighbors.add(getFieldContent(row,column + 1))
        }
        // up
        if (row != 12) {
            neighbors.add(getFieldContent(row + 1,column))
        }
        // down
        if (row != 1) {
            neighbors.add(getFieldContent(row - 1,column))
        }
        return neighbors.filter {it.color != PuyoColor.EMPTY}
    }

    // 連結数
    private fun getConnection(puyo: Puyo): List<Puyo> {
        val connected = ArrayList<Puyo>()
        if (puyo.color === PuyoColor.EMPTY) return connected
        val sameColorStack = Stack<Puyo>()
        sameColorStack.push(puyo)
        connected.add(puyo)
        while (!sameColorStack.isEmpty()) {
            for (p in getNeighborPuyo(sameColorStack.pop())) {
                if (p.color === puyo.color && !connected.contains(p)) {
                    sameColorStack.push(p)
                    connected.add(p)
                }
            }
        }
        return connected
    }

    override fun toString(): String {
        return field.joinToString("") { row -> row.map { it.color.char }.joinToString("") }
    }

    companion object {
        private val chainBonusConstant = intArrayOf(0, 0, 8, 16, 32, 64, 96, 128, 160, 192, 224, 256, 288, 320, 352, 384, 416, 448, 480, 512)
        private val colorBonusConstant = intArrayOf(0, 0, 3, 6, 12)
        private const val allClearBonus = 2100
        private fun connectionBonusConstant(connectionNum: Int): Int {
            return when {
                connectionNum <= 4 -> 0
                connectionNum <= 10 -> connectionNum - 3
                else -> 10
            }
        }
        fun from(fieldStr: String) : Field {
            val ret = Field()
            val colorString = Array(13) { i -> fieldStr.padEnd(6*13).substring(i*6, (i+1)*6) }
            ret.field = Array(13) { i -> Array(6) { j -> Puyo(i+1, j+1, getPuyoColor(colorString[i][j]))} }
            for (col in 0..5) {
                var i = 0
                while (ret.field[i++][col].color != PuyoColor.EMPTY) {
                    ret.heights[col]++
                }
            }
            return ret
        }
    }
}